### PR TITLE
Add Flint Spark index management API

### PR DESCRIPTION
**Pull request by dai-chen**
### Description

1. Draft Flint doc: https://github.com/dai-chen/sql-1/blob/add-flint-spark-api/flint/docs/index.md (put everything in one doc for now)
2. Add high level Flint Spark API `FlintSpark` and `FlintSparkSkippingIndex`
3. Add a basic Partition index `PartitionSkippingStrategy` and `FlintClient.deleteIndex()` for integration test

**TODO**

Implement Partition index building once Flint batch writer is ready

**API Example**
 
```
val flint = new FlintSpark(spark)

flint.skippingIndex()
    .onTable("alb_logs")
    .filterBy("time > 2023-04-01 00:00:00")
    .addPartitionIndex("year", "month", "day")
    .addValueListIndex("elb_status_code")
    .addBloomFilterIndex("client_ip")
    .create()
```

Please find more details in Flint doc, IT and detailed design in https://github.com/opensearch-project/sql/issues/1602.

### Issues Resolved

https://github.com/opensearch-project/sql/issues/1602
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalog.Column
+
+/**
+ * Flint Spark integration API entrypoint.
+ */
+class FlintSpark(val spark: SparkSession) {
+
+  /** Flint client for low-level index operation */
+  private val flintClient: FlintClient = {
```
Comment (penghuo) [inline flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala:27]: use FlintClientBuilder? https://github.com/opensearch-project/sql/blob/feature/flint/flint/flint-core/src/main/scala/org/opensearch/flint/core/FlintClientBuilder.java

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    }
+  }
+}
+
+object FlintSpark {
+
+  /**
+   * Flint configurations in Spark. TODO: shared with Flint data source config?
+   */
+  val FLINT_INDEX_STORE_LOCATION = "spark.flint.indexstore.location"
```
Comment (penghuo) [inline flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala:101]: QQ, should we put all the flint spark related configuration in here?

**Review by penghuo — APPROVED** [review]

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala (hunk 2 above)
Comment (dai-chen) [reply flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala:100]: yes, I think we should put all together here or a new `FlintSparkConf` class later.

File: flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+     * Add partition skipping indexed columns.
+     *
+     * @param colNames
+     *   indexed column names
+     * @return
+     *   index builder
+     */
+    def addPartitionIndex(colNames: String*): IndexBuilder = {
+      colNames
+        .map(colName => new PartitionSkippingStrategy((colName, allColumns(colName).dataType)))
```
[the remark this answers is not in the document]
Comment (dai-chen) [reply flint/flint-spark-integration/src/main/scala/org/opensearch/flint/spark/FlintSpark.scala]: Sure, add more validation in https://github.com/opensearch-project/sql/pull/1636/commits/039ef8ee9a8ebb3ba4bba0fcacc5e64207712e87. Thanks!

**Pull request merged**